src/Corax/Queries/SortingMultiMatch.cs

```diff
@@ -0,0 +1,498 @@
+                case 3: return current._comparer5;
+                case 4: return current._comparer6;
+                case 5: return current._comparer7;
+                case 6: return current._comparer8;
+                case 7: return current._comparer9;
+            }
+
+            throw new NotSupportedException("MultiMatchComparer only support up to 9 different comparisons.");
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+        private static int CompareSequence<TComparer>(ref SortingMultiMatch<TInner, TComparer1, TComparer2, TComparer3, TComparer4, TComparer5, TComparer6, TComparer7, TComparer8, TComparer9> current, int comparerIdx, UnmanagedSpan item1, UnmanagedSpan item2)
+            where TComparer : struct, IMatchComparer
+        {
+            var comparer = (TComparer)GetComparer(ref current, comparerIdx);
+            var comp1Reader = new IndexEntryReader(item1);
+            var comp2Reader = new IndexEntryReader(item2);
+
+            bool read1 = comp1Reader.Read(comparer.FieldId, out var sv1);
+            bool read2 = comp2Reader.Read(comparer.FieldId, out var sv2);
+            if (read1 && read2)
+            {
+                var result = comparer.CompareSequence(sv1, sv2);
+                int nextComparer = comparerIdx + 1;
+                if (result == 0 && nextComparer < current._totalComparers)
+                {
+                    return current._compareFuncs[nextComparer](ref current, nextComparer, item1, item2);
+                }
+                return result;
+            }
+
+            if (read1)
+                return -1;
+            return 1;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+        private static int CompareNumerical<TComparer, T>(ref SortingMultiMatch<TInner, TComparer1, TComparer2, TComparer3, TComparer4, TComparer5, TComparer6, TComparer7, TComparer8, TComparer9> current, int comparerIdx, UnmanagedSpan item1, UnmanagedSpan item2)
+            where TComparer : struct, IMatchComparer
+        {
+            var comparer = (TComparer)GetComparer(ref current, comparerIdx);
+            var comp1Reader = new IndexEntryReader(item1);
+            var comp2Reader = new IndexEntryReader(item2);
+
+            bool read1, read2;
+
+            if (typeof(T) == typeof(long) )
+            {
+                read1 = comp1Reader.Read<long>(comparer.FieldId, out var si1);
+                read2 = comp2Reader.Read<long>(comparer.FieldId, out var si2);
+                if (read1 && read2)
+                {
+                    var result = comparer.CompareNumerical<long>(si1, si2);
+                    int nextComparer = comparerIdx + 1;
+                    if (result == 0 && nextComparer < current._totalComparers)
+                    {
+                        return current._compareFuncs[nextComparer](ref current, nextComparer, item1, item2);
```

Comment:
This effectively generate a sequence of static calls that are number of comparers deep (after the first one). This can be improved upon with some boilerplate code. BUT, unless we prove this is needed with some good profiles showing that this is costly in those cases, I would go with the simple version.

Reply:
I would expect that to be pretty rare, yes.